### ISSUE TITLE
Spark 3 with Scala 2.12

### ISSUE THREAD

**Issue by LorenzBuehmann** [intel-analytics/BigDL#3070]
Hi devs!

Not an issue but a question (feel free to add the label):

Given that Spark 3 has been released now a few month ago, do you have any plans to go for it and update to Spark 3 which also needs to move to Scala 2.12?
I'm also wondering if this [Spark 3 deployment script](https://github.com/intel-analytics/BigDL/blob/master/scripts/deploy-spark3.sh) can work with the current version still being on Scala 2.11

Thanks in advance.

**Comment by jason-dai** [comment]
@LorenzBuehmann Yes, we are working on Spark 3 support with Scala 2.12, BigDL first and then Analytics Zoo (which depends on BigDL).

**Comment by Le-Zheng** [comment]
Hi LorenzBuehmann ,
The [Spark 3 deployment script](https://github.com/intel-analytics/BigDL/blob/master/scripts/deploy-spark3.sh) compiles with spark 3 and scala 2.12.  We deployed BigDL Spark3 nightly-build jar dependencies on the [Nexus repository](https://oss.sonatype.org/content/repositories/snapshots/com/intel/analytics/bigdl/dist-spark-3.0.0-scala-2.12.8-all/).  Since Spark3 is compiled with Scala 2.12,  it should work on Scala 2.12.

**Comment by mullerhai** [comment]
> Hi LorenzBuehmann ,
> The [Spark 3 deployment script](https://github.com/intel-analytics/BigDL/blob/master/scripts/deploy-spark3.sh) compiles with spark 3 and scala 2.12. We deployed BigDL Spark3 nightly-build jar dependencies on the [Nexus repository](https://oss.sonatype.org/content/repositories/snapshots/com/intel/analytics/bigdl/dist-spark-3.0.0-scala-2.12.8-all/). Since Spark3 is compiled with Scala 2.12, it should work on Scala 2.12.

HI, I want to know ,when we will release this version for bigdl build and suit with  spark 3 and scala 2.12  on normal maven public repository
thanks

**Comment by Le-Zheng** [comment]
> > Hi LorenzBuehmann ,
> > The [Spark 3 deployment script](https://github.com/intel-analytics/BigDL/blob/master/scripts/deploy-spark3.sh) compiles with spark 3 and scala 2.12. We deployed BigDL Spark3 nightly-build jar dependencies on the [Nexus repository](https://oss.sonatype.org/content/repositories/snapshots/com/intel/analytics/bigdl/dist-spark-3.0.0-scala-2.12.8-all/). Since Spark3 is compiled with Scala 2.12, it should work on Scala 2.12.
> 
> HI, I want to know ,when we will release this version for bigdl build and suit with spark 3 and scala 2.12 on normal maven public repository
> thanks

Hi, thanks for asking. We have released BigDL 0.12.1 that supports Spark 3 and scala 2.12. The normal maven repository link: https://repo1.maven.org/maven2/com/intel/analytics/bigdl/dist-spark-3.0.0-scala-2.12.10-all/0.12.1/

**Comment by LorenzBuehmann** [comment]
Hi, thank's for making the release and letting us know - good job.

Minor comment: you should also put the release on Github I think: https://github.com/intel-analytics/BigDL/releases

**Comment by JWenBin** [comment]
so 'BigDL release 0.12.1' are using TensorFlow2+ and Keras2+?

**Comment by JWenBin** [comment]
> 
> 
> @LorenzBuehmann Yes, we are working on Spark 3 support with Scala 2.12, BigDL first and then Analytics Zoo (which depends on BigDL).

How could I know the version of ‘TensorFlow/Keras’ in BigDL ?

**Comment by Le-Zheng** [comment]
> > @LorenzBuehmann Yes, we are working on Spark 3 support with Scala 2.12, BigDL first and then Analytics Zoo (which depends on BigDL).
> 
> How could I know the version of ‘TensorFlow/Keras’ in BigDL ?

Hi JWenBin,

Our project [Analytics Zoo](https://github.com/intel-analytics/analytics-zoo) supports TF/Keras/pytorch. You may use the Orca API to scale out your TensorFlow/Keras model on Spark. See https://analytics-zoo.github.io/master/#Orca/overview/
Thank you.

**Comment by JWenBin** [comment]
> 
> 
> > > @LorenzBuehmann Yes, we are working on Spark 3 support with Scala 2.12, BigDL first and then Analytics Zoo (which depends on BigDL).
> > 
> > 
> > How could I know the version of ‘TensorFlow/Keras’ in BigDL ?
> 
> Hi JWenBin,
> 
> Our project [Analytics Zoo](https://github.com/intel-analytics/analytics-zoo) supports TF/Keras/pytorch. You may use the Orca API to scale out your TensorFlow/Keras model on Spark. See https://analytics-zoo.github.io/master/#Orca/overview/
> Thank you.

我想知道BigDL 中TF和Keras的版本，你却叫我去使用你们的“analytics-zoo”，o(╯□╰)o 其实我就是想在spark下做深度的模型，所以想找个能支持TensorFlow2和Keras2的第三方实现。要是能直接在spark下构建、训练模型就更好了。

**Comment by JWenBin** [comment]
it look like BigDL is so highly similar to analytics-zoo.....

**Comment by jason-dai** [comment]
> it look like BigDL is so highly similar to analytics-zoo.....

@JWenBin BigDL does not support TF or Keras. On the other hand, Analytics Zoo does support distributed TF ad Keras (both training and inference) on Spark. See https://analytics-zoo.readthedocs.io/ for more details.

**Comment by JWenBin** [comment]
> distributed

thanks!

**Comment by JWenBin** [comment]
Analytics Zoo does not support windows: https://github.com/intel-analytics/analytics-zoo/issues/3415

**Comment by bendavidsteel** [comment]
If BigDL now supports Spark 3, does it support PySpark 3+? If so could the pip package be updated to allow pyspark>=2.4.6?

**Comment by hkvision** [comment]
Hi @bendavidsteel 
Yes, we build separate packages for pyspark 3.1.2, instead of installing from pypi, you need to download and install the packages from here: https://sourceforge.net/projects/analytics-zoo/files/ (find the folders that end with -spark3 for the nightly builds).

**Comment by bendavidsteel** [comment]
Great thanks, will use these. Is there any chance these alternative packages might eventually be available on pip or preferably conda?

**Comment by hkvision** [comment]
Hi @bendavidsteel 
We are still working on this and hopefully it will be ready soon!

**Comment by hkvision** [comment]
@bendavidsteel 

Hi, we have now released bigdl based on spark3 on pypi and you may find the instructions here: https://bigdl.readthedocs.io/en/latest/doc/UserGuide/python.html#bigdl-on-spark-3
